ocaml/utils/global.ml

```diff
@@ -13,134 +24,209 @@ module Name = struct
+end
+
+include T0
+
+let all_args t = t.visible_args @ t.hidden_args
 
 module Subst = Name.Map
 type subst = t Subst.t
 
-(* CR-someday lmaurer: Should try and make this unnecessary or at least cheap.
-   Could do it by making [Name.t] an unboxed existential so that converting from
-   [t] is the identity. Or just have [Name.t] wrap [t] and ignore [params]. *)
-let rec to_name ({ head; args; params = _ }) : Name.t =
-  { head; args = List.map arg_to_name args }
-and arg_to_name (name, value) =
-  name, to_name value
-
-let union_args args1 args2 =
-  (* [Misc.Stdlib.List.merge_map] of empty will iterate through the whole
-      other list, so special-case that out here *)
-  match args1, args2 with
-  | [], args | args, [] -> args
-  | _, _ ->
-      Misc.Stdlib.List.merge_map args1 args2
-        ~cmp:(fun (name1, _) (name2, _) -> Name.compare name1 name2)
-        ~left_only:(fun pair1 -> pair1)
-        ~right_only:(fun pair2 -> pair2)
-        ~both:
-          (fun (param, value1) (_param, value2) ->
-            Misc.fatal_errorf
-              "Duplicate arguments for %a:@ %a@ vs. %a"
-              Name.print param
-              print value1
-              print value2)
-
-let rec subst ({ head; args; params } as param) (s : subst) =
-  match Subst.find_opt (to_name param) s with
-  | Some rhs -> rhs
-  | None ->
-      let matching_params, non_matching_params =
-        List.partition_map
-          (fun (name, value) ->
-              match Subst.find_opt (to_name value) s with
-              | Some rhs -> Left (name, rhs)
-              | None -> Right (name, value))
-          params
-      in
-      let args = List.map (fun pair -> subst_pair pair s) args in
-      let params =
-        List.map (fun pair -> subst_pair pair s) non_matching_params
-      in
-      let args = union_args args matching_params in
-      { head; args; params }
-and subst_pair (name, value) s =
-  name, subst value s
-
-let check s params =
+let rec subst0 (t : t) (s : subst) ~changed =
+  match Subst.find_opt (to_name t) s with
+  | Some rhs -> changed := true; rhs
+  | None -> subst0_inside t s ~changed
+and subst0_inside { head; visible_args; hidden_args } s ~changed =
+  let matching_hidden_args, non_matching_hidden_args =
+    List.partition_map
+      (fun ((name, value) as pair) ->
+          match Subst.find_opt (to_name value) s with
+          | Some rhs -> changed := true; Left (name, rhs)
+          | None -> Right pair)
+      hidden_args
+  in
+  let visible_args = subst0_alist visible_args s ~changed in
+  let hidden_args = subst0_alist non_matching_hidden_args s ~changed in
+  let visible_args =
+    List.merge compare_arg_name visible_args matching_hidden_args
+  in
+  create head visible_args ~hidden_args
+and subst0_alist l s ~changed =
+  List.map (fun (name, value) -> name, subst0 value s ~changed) l
+
+let subst t s =
+  let changed = ref false in
+  let new_t = subst0 t s ~changed in
+  if !changed then new_t else t
```

Comment:
Best to return `!changed` rather than force user to use `==`, whose definition is imprecise on a immutable values.